ocaml-lsp-server/src/compl.ml

```diff
@@ -119,62 +115,124 @@ let range_prefix (lsp_position : Position.t) prefix : Range.t =
          provided by merlin. *)
     ~sortText:(Printf.sprintf "%04d" index)
-    ~data:compl_info ?textEdit ()
+    ~data:compl_params ~textEdit ()
 
-let complete doc lsp_position =
-  let position = Position.logical lsp_position in
-
-  let prefix =
-    prefix_of_position ~short_path:false (Document.source doc) position
-  in
-  Log.log ~section:"debug" (fun () ->
-      Log.msg "completion prefix" [ ("prefix", `String prefix) ]);
-
-  Document.with_pipeline_exn doc @@ fun pipeline ->
-  let completion =
-    let complete =
-      Query_protocol.Complete_prefix (prefix, position, [], false, true)
-    in
-    Query_commands.dispatch pipeline complete
+let complete_prefix doc prefix pos =
+  let open Fiber.O in
+  let position = Position.logical pos in
+  let+ (completion : Query_protocol.completions) =
+    (* TODO: can't we use [Document.dispatch_exn] instead of
+       [Document.with_pipeline_exn] for conciseness? *)
+    Document.with_pipeline_exn doc (fun pipeline ->
+        let complete =
+          Query_protocol.Complete_prefix (prefix, position, [], false, true)
+        in
+        Query_commands.dispatch pipeline complete)
   in
-  let short_range =
-    range_prefix lsp_position
+  let range =
+    range_prefix pos
       (prefix_of_position ~short_path:true (Document.source doc) position)
   in
-  let items =
-    completion.entries |> List.map ~f:(fun entry -> (short_range, entry))
-  in
-  let items =
+  let completion_entries =
     match completion.context with
-    | `Unknown -> items
+    | `Unknown -> completion.entries
     | `Application { Query_protocol.Compl.labels; argument_type = _ } ->
-      items
+      completion.entries
       @ List.map labels ~f:(fun (name, typ) ->
-            ( short_range
-            , { Query_protocol.Compl.name
-              ; kind = `Label
-              ; desc = typ
-              ; info = ""
-              ; deprecated = false (* TODO this is wrong *)
-              } ))
+            { Query_protocol.Compl.name
+            ; kind = `Label
+            ; desc = typ
+            ; info = ""
+            ; deprecated = false (* TODO this is wrong *)
+            })
   in
-  let textDocument = TextDocumentIdentifier.create ~uri:(Document.uri doc) in
-  let compl_info =
-    CompletionParams.create ~textDocument ~position:lsp_position ()
+  (* we need to json-ify completion params to put them in completion item's
+     [data] field to keep it across [textDocument/completion] and the following
+     [completionItem/resolve] requests *)
+  let compl_params =
+    let textDocument = TextDocumentIdentifier.create ~uri:(Document.uri doc) in
+    CompletionParams.create ~textDocument ~position:pos ()
     |> CompletionParams.yojson_of_t
   in
-  let items = List.mapi ~f:(item ~compl_info) items in
-  `CompletionList { CompletionList.isIncomplete = false; items }
+  List.mapi completion_entries
+    ~f:(completionItem_of_completion_entry ~range ~compl_params)
+
+let construct doc position =
```

Comment:
The return type for this function is needlessly informative. Let's go over the three cases:

* There's an exception we don't know how to handle. We can't do anything except bubble it up to the client. No use in informing the caller.
* We can provide completion candidates - okay, let's return them.
* There are no completion candidates (for whatever reason, no hole, nothing to suggest) - an empty list is all the caller needs to know.

Thus I expect function should return `CompletionItem.t list Fiber.t`.